src/clr-angular/data/datagrid/render/main-renderer.ts

```diff
@@ -100,6 +109,15 @@ export class DatagridMainRenderer<T = any> implements AfterContentInit, AfterVie
     }
   }
 
+  private setupColumns() {
+    this.columnsService.columns = this.headers.toArray().map(() => new BehaviorSubject<DatagridColumnState>({}));
```

Comment:
Dynamic columns with `*ngIf` or changing `*ngFor` will lose all information of the current column state. I'm pretty sure in the current version this would be pretty bad anyway, but I figured I'd point it out if anyone has a good idea. I don't think it's critical, except maybe for columns that the user manually resized.

Reply:
I had thought about this, but the current behavior will lose the column state too. We'd have to somehow know to cache that state and carry it over, something I can't see a good solution for without the app passing some configuration. Something for another day I guess.